processing/utils/diffing.ts

```diff
@@ -0,0 +1,118 @@
+import { PrismaClient } from '@prisma/client'
+import type { Topic } from '../topics.const'
+import { $ } from 'bun'
+
+const backupTableIdentifier = (table: string) => `backup."${table}"`
+const diffTableIdentifier = (table: string) => `public."${table}_diff"`
+const tableIdentifier = (table: string) => `public."${table}"`
+export async function getTopicTables(topic: Topic) {
+  return $`lua /processing/utils/TableNames.lua ${topic}`
+    .text()
+    .then((tables) => tables.split('\n'))
+}
+
+const prisma = new PrismaClient()
+
+export async function createBackupSchema() {
+  prisma.$executeRaw`CREATE SCHEMA IF NOT EXISTS backup`
+}
+export async function getSchemaTables(schema: string) {
+  const rows: { table_name: string }[] = await prisma.$queryRaw`
+    SELECT table_name
+    FROM information_schema.tables
+    WHERE table_schema = ${schema}
+    AND table_type = 'BASE TABLE';`
+  return new Set(rows.map(({ table_name }) => table_name))
+}
+export async function backupTable(table: string) {
+  const tableId = tableIdentifier(table)
+  const backupTableId = backupTableIdentifier(table)
+  await prisma.$queryRawUnsafe(`DROP TABLE IF EXISTS ${backupTableId}`)
+  return prisma.$queryRawUnsafe(`CREATE TABLE ${backupTableId} AS TABLE ${tableId}`)
+}
+
+export async function dropDiffTable(table: string) {
+  const diffTableId = diffTableIdentifier(table)
+  return prisma.$executeRawUnsafe(`DROP TABLE ${diffTableId}`)
+}
+
+export async function computeDiff(table: string) {
+  const tableId = tableIdentifier(table)
+  const backupTableId = backupTableIdentifier(table)
+  const diffTableId = diffTableIdentifier(table)
+  const joinedTableId = `backup."${table}_joined"`
+  const changeTypes = {
```

Comment:
Future: Jetzt wo wir in JS sind, könnten wir auf Tools wie https://www.npmjs.com/package/git-diff zurückgreifen, oder? Dann kann man das im UI direkt mit so einem Diff-Highlightin darstellen mit sowas wie https://highlightjs.readthedocs.io/en/latest/supported-languages.html ?

Reply:
Nein, das Diffing wird komplett in Postgres aus dem `JSONB` columns ausgeführt. Wenn wir das package benutzen wollen, müssten wir es als string aus postgres in JS laden und dann dort vergleichen und zurück schreiben. Das wäre auf jeden Fall deutlich langsamer.